orangecontrib/prototypes/widgets/owattributes.py

```diff
@@ -0,0 +1,256 @@
+import locale
+from collections import defaultdict
+
+import numpy as np
+import scipy.stats as stats
+from AnyQt.QtCore import Qt, QSize, QRectF, QAbstractTableModel
+from AnyQt.QtGui import QStandardItemModel, QStandardItem
+from AnyQt.QtWidgets import QHeaderView, QStyledItemDelegate, QGraphicsScene
+
+from Orange.data import Table
+from Orange.widgets import widget, gui
+from Orange.widgets.settings import Setting
+from Orange.widgets.utils.signals import Input
+from orangecontrib.prototypes.widgets.utils.histogram import Histogram
+
+
+class Stats:
+    def __init__(self):
+        self.__cache = defaultdict(dict)
+
+    # TODO implement a memoize helper function that only remembers attribute
+    def first_moment(self, attribute, data):
+        if attribute in self.__cache['first']:
+            return self.__cache['first'][attribute]
+
+        if attribute.is_discrete:
+            idx = int(stats.mode(data).mode[0])
+            result = data[idx]
+        elif attribute.is_continuous:
+            result = np.mean(data)
+
+        self.__cache['first'][attribute] = result
+        return result
+
+    def second_moment(self, attribute, data):
+        if attribute in self.__cache['second']:
+            return self.__cache['second'][attribute]
+
+        if attribute.is_discrete:
+            result = stats.entropy(data)
+        elif attribute.is_continuous:
+            result = np.var(data)
+
+        self.__cache['second'][attribute] = result
+        return result
+
+    def valid(self, attribute, data):
+        if attribute in self.__cache['valid']:
+            return self.__cache['valid'][attribute]
+
+        result = int((~np.isnan(data)).sum())
```

Comment:
Please use our own implementation of `countnans` (`Orange/statistics/util.py`) which can also handle sparse data sets.